### PR TITLE
Issue 243 - Introduce AppSetup for Switching Environments

### PR DESCRIPTION
**Pull request by ahmed-shehata**
Issue #243

### CODE REVIEW

File: AtlasDemo/AtlasDemo/AppDelegate.swift [hunk 1]
```diff
@@ -4,63 +4,16 @@
 
 import UIKit
 
-import AtlasSDK
-import AtlasUI
-import AtlasMockAPI
-
-var AtlasCheckoutInstance: AtlasCheckout? // swiftlint:disable:this variable_name
-
+var AppSetupInstance = AppSetup() // swiftlint:disable:this variable_name
```
Comment (bonkey) [inline AtlasDemo/AtlasDemo/AppDelegate.swift]: Just make `AppSetup.sharedInstance`. It's much better than free global instance. In the end it's exactly same thing.

File: AtlasDemo/AtlasDemo/AppSetup.swift [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+//  Copyright © 2016 Zalando SE. All rights reserved.
+//
+
+import Foundation
+import AtlasSDK
+import AtlasUI
+import AtlasMockAPI
+
+class AppSetup {
+    var checkout: AtlasCheckout? // swiftlint:disable:this variable_name
```
Comment (bonkey) [inline AtlasDemo/AtlasDemo/AppSetup.swift]: What's wrong, that you need to disable swiftlint?

File: AtlasDemo/AtlasDemo/AppSetup.swift [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+        #endif
+    }
+
+    private func prepareApp() {
+        if AtlasMockAPI.hasMockedAPIStarted {
+            Atlas.logoutCustomer()
+        }
+    }
+
+    private func setAppOptions(opts: Options, completion: (() -> Void)? = nil) {
```
Comment (bonkey) [inline AtlasDemo/AtlasDemo/AppSetup.swift]: Just name it `configure`. It's not doing anything else but calling another configure

File: AtlasDemo/AtlasDemo/AppSetup.swift [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+        AtlasCheckout.configure(opts) { result in
+            if case let .success(checkout) = result {
+                self.checkout = checkout
+                if let completion = completion {
```
Comment (bonkey) [inline AtlasDemo/AtlasDemo/AppSetup.swift]: just `completion?()` without any ifs

File: AtlasDemo/AtlasDemo/AppSetup.swift [hunk 5]
```diff
@@ -0,0 +1,68 @@
...
+            useSandbox: useSandbox, interfaceLanguage: "en_DE")
+
+        if AtlasMockAPI.hasMockedAPIStarted {
+            opts = Options(basedOn: opts, configurationURL: AtlasMockAPI.endpointURL(forPath: "/config"))
+        }
+
+        return opts
+    }
+
+    private func prepareMockAPI() {
```
Comment (bonkey) [inline AtlasDemo/AtlasDemo/AppSetup.swift]: Put all funcs in the order of calling them.

File: AtlasDemo/AtlasDemo/CatalogViewController.swift [hunk 6]
```diff
@@ -51,6 +53,17 @@ class CatalogViewController: UIViewController, UICollectionViewDelegate, UIColle
...
     }
 
+    @IBAction func serverSwitchTapped(sender: AnyObject) {
+        if let serverSwitch = sender as? UISwitch {
+            switch serverSwitch.on {
+            case true:
+                AppSetupInstance.switchEnvironment(useSandbox: false)
+            default:
+                AppSetupInstance.switchEnvironment(useSandbox: true)
+            }
```
Comment (bonkey) [inline AtlasDemo/AtlasDemo/CatalogViewController.swift]: @hani-ibrahim You and your spoiled switch-based-programming!

@ahmed-shehata Just as it is: `AppSetupInstance.switchEnvironment(useSandbox: !serverSwitch.on)`

File: AtlasDemo/AtlasDemo/CatalogViewController.swift [hunk 7]
```diff
@@ -55,14 +55,8 @@ class CatalogViewController: UIViewController, UICollectionViewDelegate, UIColle
 
     @IBAction func serverSwitchTapped(sender: AnyObject) {
```
Comment (bonkey) [inline AtlasDemo/AtlasDemo/CatalogViewController.swift]: Just `...switch: UISwitch)`

**Comment by bonkey** [comment]
👍

**Comment by ahmed-shehata** [comment]
👍

**Pull request merged**